### DISCUSSION TITLE
Compiling 'oc'

### DISCUSSION THREAD

**Discussion by darren-oxford** [okd-project/okd#1298] [General]
The OpenShift Client binaries for MacOS arm64 don't work, either the OKD or OpenShift supplied ones. I therefore tried to compile my own, this works but 'oc version --client' reports v4.2.0-alpha.0-1548-g7075089 when I compile the latest master brach.

Obviously there is some secret step missing in the official build process that sets the correct version number, anyone know what that would be?

**Comment by nate-duke** [comment]
@darren-oxford did you try pointing at the tagged release? e.g.
https://github.com/openshift/okd/tree/4.11.0-0.okd-2022-07-29-154152

Hrm, [the oc repository](https://github.com/openshift/oc) is organized quite differently so i guess it's no so clear. Now I'm wondering how to go about this myself.

**Reply by darren-oxford** [reply]
Thanks @nate-duke, I tried cloning the release-4.11  from https://github.com/openshift/oc.git and it is working, just that version is reported as v4.2.0-alpha.x. I think it is building 4.11, so should be usable but would be good to figure out how to get the build working properly to report the correct version.

**Reply by darren-oxford** [reply]
OK, it seems to be down to the way the repository is organised.

It is a rather manual process but I have achieved it as follows ...

- git clone -b https://github.com/openshift/oc.git (to get everything)
- git tag -l (to view tags, select the one you want, e.g openshift-clients-4.11.0-202204020828)
- git checkout tags/openshift-clients-4.11.0-202204020828
- export OS_GIT_VERSION="4.11.0-202204020828"
- CGO_CFLAGS="-I/opt/homebrew/opt/heimdal/include" make oc

Note the CGO_CFLAGS environment variable is only required on a Mac as heimdal needs to be installed using homebrew to complete the compilation.

**Reply by nate-duke** [reply]
I'm sure this is all clear (or at least inferrable) somewhere in Red Hat's Build System but it's nice to have it here for reference and for future folks. Thanks for following up!

**Comment by vrutkovs** [comment]
Check out https://github.com/okd-project/okd-centos9-rebuild/blob/main/buildconfigs/10-oc.yaml for an image with `oc` and https://github.com/okd-project/okd-centos9-rebuild/blob/main/buildconfigs/20-cli-artifacts.yaml for image with platform-specific builds
